source/val/validate_memory.cpp

```diff
@@ -712,6 +708,115 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
               "parameters";
   }
 
+  if (_.HasCapability(SpvCapabilityShader)) {
+    // Don't allow variables containing 16-bit elements without the appropriate
+    // capabilities.
+    if ((!_.HasCapability(SpvCapabilityInt16) &&
+         _.ContainsSizedIntOrFloatType(value_id, SpvOpTypeInt, 16)) ||
+        (!_.HasCapability(SpvCapabilityFloat16) &&
+         _.ContainsSizedIntOrFloatType(value_id, SpvOpTypeFloat, 16))) {
+      auto underlying_type = value_type;
+      while (underlying_type->opcode() == SpvOpTypePointer) {
+        storage_class = underlying_type->GetOperandAs<SpvStorageClass>(1u);
+        underlying_type =
+            _.FindDef(underlying_type->GetOperandAs<uint32_t>(2u));
+      }
+      bool storage_class_ok = true;
+      std::string sc_name = _.grammar().lookupOperandName(
+          SPV_OPERAND_TYPE_STORAGE_CLASS, storage_class);
+      switch (storage_class) {
+        case SpvStorageClassStorageBuffer:
+        case SpvStorageClassPhysicalStorageBufferEXT:
+          if (!_.HasCapability(SpvCapabilityStorageBuffer16BitAccess)) {
+            storage_class_ok = false;
+          }
+          break;
+        case SpvStorageClassUniform:
+          if (!_.HasCapability(
+                  SpvCapabilityUniformAndStorageBuffer16BitAccess)) {
+            if (value_type->opcode() == SpvOpTypeArray ||
```

Comment:
This looks wrong.  You're changing the value type from enclosing struct to the contained array/RTArray. Then checking for the decoration on the array type.  But the BufferBlock decoration goes on the struct.
Seems like this whole "if" clause should be removed.

Reply:
Right, I think it should `underlying_type` instead. That starts as a copy of `value_type`, which is the first pointed to type (i.e. the variable's allocated type). To find the Block/BufferBlock decorated struct we have to resolve the potential array-ness of the variable. I will add a few tests for this.